Add some basic preferences for the monaco editor

epatpol opened the pull request:
This adds the possibility of changing the tabSize and showing/hiding the
lineNumbers by modifying the .theia/settings.json file.

Signed-off-by: Patrick-Jeffrey Pollo Guilbert <patrick.pollo.guilbert@ericsson.com>

epatpol commented:
This is based on @akosyakov's work for the filesystemwatcher. It's still a WIP, so don't merge yet, but feel free to add comments.

akosyakov commented:
so it should be:
```ts
export interface EditorConfiguration {
    'editor.tabSize': number,
    'editor.lineNumbers': 'on' | 'off'
}
```

epatpol commented:
That makes sense, that's what I thought after pushing the patch also (putting more generic preferences in editor and more specific preferences in the monaco/X editor implementation), I'll change it asap.

akosyakov commented on src/monaco/browser/monaco-editor-provider.ts:
```diff
@@ -40,16 +43,27 @@ export class MonacoEditorProvider {
                     folding: true,
                     theme: 'vs-dark',
-                    readOnly: model.readOnly
+                    readOnly: model.readOnly,
```
don't forget to properly initialize the editor initially

akosyakov commented on src/monaco/browser/monaco-editor-provider.ts:
```diff
@@ -40,16 +43,27 @@ export class MonacoEditorProvider {
+
+            this.monacoPrefs.onPreferenceChanged(e => {
+                if (e.preferenceName === 'monaco.tabSize') {
+                    editor.getControl().getModel().updateOptions({ tabSize: <number>e.newValue });
```
we should try to find a way to get rid of type casting, but separately

epatpol commented:
Changed it to be more generic, and added the whitespaces option (thx @akosyakov)

working example:


![screencast_2017-07-10_04-29-52 pm](https://user-images.githubusercontent.com/2881761/28038571-6494adaa-658d-11e7-90a3-106ce71a74d5.gif)

akosyakov commented on src/monaco/browser/monaco-editor-provider.ts:
```diff
@@ -25,7 +27,8 @@ export class MonacoEditorProvider {
         @inject(MonacoWorkspace) protected readonly workspace: MonacoWorkspace,
-        @inject(MonacoCommandServiceFactory) protected readonly commandServiceFactory: MonacoCommandServiceFactory
+        @inject(MonacoCommandServiceFactory) protected readonly commandServiceFactory: MonacoCommandServiceFactory,
+        @inject(EditorPreferences) protected readonly monacoPrefs: EditorPreferences,
```
`editorPreferneces` or `preferences`

akosyakov commented on src/editor/browser/index.ts:
```diff
@@ -13,3 +13,4 @@ export * from './editor-command';
 export * from './editor-menu';
 export * from './editor-keybinding';
 export * from './editor-frontend-module';
+export * from './editor-preferences';
```
Where `editor-preferences` itself? I don't see it in the commit

akosyakov commented on src/monaco/browser/monaco-editor-provider.ts:
```diff
@@ -48,13 +51,33 @@ export class MonacoEditorProvider {
             );
-            editor.onDispose(() => reference.dispose());
 
+            this.monacoPrefs.onPreferenceChanged(e => {
```
you should dispose a listener if the editor is disposed

akosyakov replied:
It is not enough listening to changes, the editor should be initialized initially properly.

epatpol replied:
Oh right true that, right now the prefs would only load on settings.change

akosyakov commented on src/monaco/browser/monaco-frontend-module.ts:
```diff
@@ -20,13 +20,16 @@ import { MonacoEditorService } from "./monaco-editor-service";
 import { MonacoModelResolver } from "./monaco-model-resolver";
 import { MonacoContextMenuService } from "./monaco-context-menu";
 import { MonacoCommandService, MonacoCommandServiceFactory } from './monaco-command-service';
+import { bindEditorPreferences } from '../../editor/browser/editor-preferences'
```
it should be bound in the editor extension, the monaco extension assume that it is bound.

akosyakov commented:
only wonder what happens if the bogus value is entered

epatpol commented:
Hmmm true that, it should fall back to the default value but right now nothing happens.

akosyakov commented:
@epatpol just file a bug

epatpol commented:
Alright then I'll merge this one and open a bug thanks @akosyakov
